proposals/predict_ml_model_probabilities.json

```
@@ -0,0 +1,45 @@
+{
+    "id": "predict_ml_model_probabilities",
+    "summary": "Predict class probabilities using a ML model",
+    "description": "Applies a machine learning model to an array and predicts (class) probabilities for them.",
+    "categories": [
+        "machine learning",
+        "reducer"
+    ],
+    "experimental": true,
+    "parameters": [
+        {
+            "name": "data",
+            "description": "An array of numbers.",
+            "schema": {
+                "type": "array",
+                "items": {
+                    "type": [
+                        "number",
+                        "null"
+                    ]
+                }
+            }
+        },
+        {
+            "name": "model",
+            "description": "A ML model that can be trained with one of the ML processes such as ``fit_regr_random_forest()``.",
+            "schema": {
+                "type": "object",
+                "subtype": "ml-model"
+            }
+        }
+    ],
+    "returns": {
+        "description": "The predicted (class) probabilities. Returns `null` if any of the given values in the array is a no-data value.",
```

```suggestion
        "description": "The predicted class probabilities. Returns `null` if any of the given values in the array is a no-data value.",
```

I'm not sure about the `Returns null if any ...`:  that depends on the capability of the ML model to handle null/nodata I think

---

This is actually tricky - some thoughts:
- E.g. in deep learning if I've anticipated missing data, I could have used `torch.nan_to_num` to replace nan with a carefully chosen value in both training and inference. If I then export that model with e.g. ONNX, that transformation will be baked in and my model will know how to deal with missing data. I'm not sure how every other frameworks handles this (e.g. sklearn or xgboost) - but it's super important that this value is the same during inference as it was during training!
- If the model handles nans, and the `predict_ml_model` process just ignores nan-values and lets it through, then I get the correct behaviour. 
- However, if it fails on nan-values, then the only way the user can fix this is to replace nan-values with the correct value in an extra openeo process before. To do that I need to know what that value was during training - this might not be easily available!
- If the model doesn't handle nans (either they didn't add nan-handling to inference code, or just didn't train on samples with nans at all, etc.), then what would happen really depends on the framework. It might just crash, or it might run through, with the nan values subtly impacting the predictions. 
- some options I can think of:
	- Make this a parameter of the process (`fail_on_nan` or similar), fail by default and allow overriding if the user knows that their model can handle nans out of the box or is willing to throw the dice.
	- Add information about what to replace nan values with to the `ml-model` STAC extension and use that if available

I think as a user my preference would be for this process to assume that the model has already been constructed to handle nans correctly in both training and inference and therefore doesn't try to interfere.

---

If I understand correctly, you agree that it would be better to drop 
> Returns `null` if any of the given values in the array is a no-data value.

from the general description of `predict_..._probabilities` ?

---

> If I understand correctly, you agree that it would be better to drop
> 
> > Returns `null` if any of the given values in the array is a no-data value.
> 
> from the general description of `predict_..._probabilities` ?

Yeah, exactly!